src/main/scala/org/constellation/snapshotstreaming/storage/FileBasedLastIncrementalCurrencySnapshotStorage.scala

```diff
@@ -0,0 +1,40 @@
+package org.constellation.snapshotstreaming.storage
+
+import cats.effect.std.Semaphore
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import org.tessellation.security._
+import fs2.io.file._
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import org.tessellation.currency.schema.currency.CurrencyIncrementalSnapshot
+import org.tessellation.currency.schema.currency.CurrencySnapshotInfo
+import org.tessellation.currency.schema.currency.CurrencySnapshotStateProof
+import org.tessellation.merkletree.StateProofValidator
+
+class FileBasedLastIncrementalCurrencySnapshotStorage[F[_]: Async: Hasher](semaphore: Semaphore[F], path: Path)(implicit
+  codec: Codec[SnapshotWithState[CurrencyIncrementalSnapshot, CurrencySnapshotInfo]]
+) extends BaseLastIncrementalSnapshotStorage[F, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotInfo](semaphore, path) {
+
+  def validateStateProof(snapshot: Hashed[CurrencyIncrementalSnapshot], state: CurrencySnapshotInfo): F[Unit] =
+    StateProofValidator
+      .validate(snapshot, state)
+      .map(_.isValid)
+      .flatMap(new Throwable("State proof doesn't match!").raiseError[F, Unit].unlessA)
```

Comment:
Can we simply to this? The invalid already has a similar message and it also provides the ordinal and hash
```suggestion
      .flatMap(Async[F].fromValidated)
```